Feature: events strategy evaluation

agataben opened the pull request:
#62 is addressed
Two evaluation strategy are available:
- `strategy='flags'`: anomalies are identified at the data-point level, based on the anomaly flag associated
- `strategy='events'`: anomalies are identified at the event level, i.e. an anomalous event represents a group of anomalous data points.
At the moment, only the point granularity is implemented for the events strategy. 
About the series granularity: with this granularity, the evaluation does not distinguish between the strategies (as already discussed)

agataben committed:
Set the structure for implementing the evaluation strategy by events

New functions "_variable_eval_with_events_strategy()", "_point_eval_with_events_strategy()",
"_series_eval_with_events_strategy()" aim to the evaluation with events strategy with the 3 supported
granularities.

The use of functions above inside "evaluate()" has been established.

agataben committed:
Implement "_point_eval_with_events_strategy()" without breakdown

The implementation of the above function includes adding "_get_anomalous_events()"

agataben committed:
Implement "_get_anomalous_events()"

This function counts the total number of anomalous events in the series

agataben committed:
Add test on "_point_eval_with_events_strategy()"

agataben committed:
Fix errors on "_point_eval_with_events_strategy()"

agataben committed:
Add the breakdown

agataben committed:
Change the function "_get_anomalous_events" in “_count_anomalous_events“

Now this function counts the number of anomalies inserted for each type

agataben committed:
Complete the breakdown

agataben committed:
Fix a zero division

agataben committed:
Fix error on evaluation with events strategy and point granularity

agataben committed:
Add NotImplemented for evaluation with events strategy and variable granularity

agataben committed:
Add test on evaluation with point granularity, events strategy and breakdown

agataben committed:
Raise NotImplementedError for evaluation with variable granularity and events strategy

sarusso commented:
In the `evaluators.py` there is a bug if not requiring the breakdown. Around line 333, this block:

```
    for key in inserted_events_by_type.keys():
        breakdown_info[key + '_true_positives_rate'] = breakdown_info[key + '_true_positives_count']/inserted_events_by_type[key]
    if breakdown:
        return evaluation_result | breakdown_info
```

has to become this:

```
    if breakdown:
        for key in inserted_events_by_type.keys():
            breakdown_info[key + '_true_positives_rate'] = breakdown_info[key + '_true_positives_count']/inserted_events_by_type[key]
        return evaluation_result | breakdown_info
```

sarusso commented:
Besides the above mini-bug, it does not seem to work with "real" data.

For example, the following code:

```
import numpy as np
import random as rnd
rnd.seed(0)
np.random.seed(0)

from ats.dataset_generators import HumiTempDatasetGenerator
generator = HumiTempDatasetGenerator(sampling_interval='60m')
evaluation_dataset = generator.generate( 
    n_series = 5,
    effects = ['noise', 'clouds'],
    anomalies = ['spike_mv', 'step_mv'],
    time_span = '90D',
    max_anomalies_per_series = 3,
    anomalies_ratio = 1.0,
    auto_repeat_anomalies=True,
)

anomaly_detectors = {}
from ats.anomaly_detectors.naive.minmax import MinMaxAnomalyDetector
anomaly_detectors['minmax'] = MinMaxAnomalyDetector()
from ats.anomaly_detectors.stat.robust import NHARAnomalyDetector
anomaly_detectors['nhar'] = NHARAnomalyDetector()

from ats.evaluators import Evaluator
evaluator = Evaluator(evaluation_dataset)
evaluation = evaluator.evaluate(anomaly_detectors,
                                granularity='point',
                                strategy='events',
                                breakdown=False)
print(evaluation)
```

...gives me the following output:

```
{'minmax': {'true_positives_count': 1,
  'true_positives_rate': 0.06666666666666667,
  'false_positives_count': 0,
  'false_positives_ratio': 0.0},
 'nhar': {'true_positives_count': 0,
  'true_positives_rate': 0.0,
  'false_positives_count': 0,
  'false_positives_ratio': 0.0}}
```

Which does not make much sense.

agataben committed:
Fix bugs

About "_point_eval_with_events_strategy()"
- The event evaluation was not taking into account not detected anomalies. The "breakdown_info" dictionary had only
detected anomalies as keys; the for cycle for calculating the anomaly ratio (anomaly_counts/inserted_anomalies)
was done on keys of "inserted_events_by_type", so a key error was raised for not detected anomalies.

- The dataframe was filtrated using the column name "_anomaly" to get the model flags but the "apply()" method of the model
adds the column "anomaly". Probably this is the reason why the evaluation on models like NHARAnomalyDetector was giving
0.0 everywhere.

agataben commented:
About the first comment: I think that the real bug was not that. I was not taking into account not detected anomalies and this was raising a key error when calculating the true positives ratio for each anomaly (because I was cycling on  keys of a dictionary containing all the inserted anomalies). 
About the second comment: yes, the result is strange. Probably it was due to the fact that I was filtering the dataframe using "_anomaly" instead of "anomaly". Now it should work. I have tried also with the breakdown:
```
import numpy as np
import random as rnd
rnd.seed(0)
np.random.seed(0)

from ats.dataset_generators import HumiTempDatasetGenerator
generator = HumiTempDatasetGenerator(sampling_interval='60m')
evaluation_dataset = generator.generate( 
    n_series = 5,
    effects = ['noise', 'clouds'],
    anomalies = ['spike_mv', 'step_mv'],
    time_span = '90D',
    max_anomalies_per_series = 3,
    anomalies_ratio = 1.0,
    auto_repeat_anomalies=True,
)

anomaly_detectors = {}
from ats.anomaly_detectors.naive.minmax import MinMaxAnomalyDetector
from ats.anomaly_detectors.stat.robust import NHARAnomalyDetector
from ats.anomaly_detectors.stat.periodic_average import PeriodicAverageAnomalyDetector
anomaly_detectors['minmax'] = MinMaxAnomalyDetector()
anomaly_detectors['nhar'] = NHARAnomalyDetector()
anomaly_detectors['p_average'] = PeriodicAverageAnomalyDetector()

from ats.evaluators import Evaluator
evaluator = Evaluator(evaluation_dataset)
evaluation = evaluator.evaluate(anomaly_detectors, granularity='point',strategy='events',breakdown=True)
```
Results:
```
evaluation['minmax'] = {
true_positives_count: 1 
true_positives_rate: 0.06666666666666667 
false_positives_count: 0 
false_positives_ratio: 0.0 
step_mv_true_positives_count: 0 
step_mv_true_positives_rate: 0.0 
spike_mv_true_positives_count: 1 
spike_mv_true_positives_rate: 0.125 
}
```
```
evaluation['p_average'] = {
true_positives_count: 8 
true_positives_rate: 0.7 
false_positives_count: 11 
false_positives_ratio: 0.0010304449648711944 
step_mv_true_positives_count: 5 
spike_mv_true_positives_count: 3 
step_mv_true_positives_rate: 0.7333333333333333 
spike_mv_true_positives_rate: 0.625 
}
```
```
evaluation['nhar'] = {
true_positives_count: 1 
true_positives_rate: 0.06666666666666667 
false_positives_count: 4 
false_positives_ratio: 0.0003703703703703704 
step_mv_true_positives_count: 0 
step_mv_true_positives_rate: 0.0 
spike_mv_true_positives_count: 1 
spike_mv_true_positives_rate: 0.125 
}
```

sarusso commented:
I am not sure this is correct with respect to false positives. Why in the "minmax" there are no false positives at all? Does not seem right...

agataben commented:
Yes, you are right. I'am working on it

agataben committed:
Add a test to detect incorrect false positive counting

agataben committed:
Delete the instance of "PeriodicAverageAnomalyDetector()" in "test_correct_counting_false_positives_with_events_strategy()"

This test checks if true positives and false positives are correctly counted using a dataset with only 1 series.
In this case, the ratio between true positives (false positives) count and true positives rate (false positives ratio)
must give the total number of inserted anomalies (lenght of the series).
Tests pass for "MinMaxAnomalyDetector()" and "NHARAnomalyDetector()" but not for the "PeriodicAverageAnomalyDetector()"
because the output of the model is a shorter series than the input one.

agataben committed:
Fix the problem with the periodic average model

Now the false positives ratio is computed as the ratio between false positives counts and the length
of the column 'anomaly_labels', which is equal to the length of the series of the dataset.
This was done because the function "_point_eval_with_events_strategy()" takes in input the flagged series
(output of the "apply()" method of the model) that for the periodic average model has a shorter length than the
original series of the dataset.

agataben commented:
I have done a test to check the correct counting of true positives and false positives.
In this test, the evaluation dataset has only 1 series and, for each evaluated model, I check if:
- true_positives_count/true_positives_rate = n. inserted anomalies (1)
- false_positives_count/false_positives_ratio = length of the series (2)

(**The equations above are true only for 1 series dataset**)
The models taken into account in the test are:
1) `MinMaxAnomalyDetector()`
2) `NHARAnomalyDetector()`
3) `PeriodicAverageAnomalyDetector()`

The evaluation result of all the models is in agreement with (1); only the periodic average disagrees with (2). 
But this should be normal, because the model output is a shorter series than the input one. 
To fix this problem, now  the false positives ratio is computed as the ratio between false positives counts and the length of the column 'anomaly_labels', which is equal to the length of the series of the dataset.
Now the result of the evaluation is:
```
evaluation['minmax'] = {
true_positives_count: 11 
true_positives_rate: 0.9333333333333332 
false_positives_count: 9 
false_positives_ratio: 0.0008333333333333335 
step_mv_true_positives_count: 10 
step_mv_true_positives_rate: 1.7333333333333332 
spike_mv_true_positives_count: 1 
spike_mv_true_positives_rate: 0.125 
}
```
```
evaluation['nhar'] = {
true_positives_count: 14 
true_positives_rate: 1.1 
false_positives_count: 58 
false_positives_ratio: 0.005370370370370371 
step_mv_true_positives_count: 13 
step_mv_true_positives_rate: 1.6666666666666667 
spike_mv_true_positives_count: 1 
spike_mv_true_positives_rate: 0.125 
}
```
```
evaluation['p_avg'] = {
true_positives_count: 165 
true_positives_rate: 13.4 
false_positives_count: 1597 
false_positives_ratio: 0.14787037037037037 
step_mv_true_positives_count: 162 
spike_mv_true_positives_count: 3 
step_mv_true_positives_rate: 23.466666666666665 
spike_mv_true_positives_rate: 0.625 
}
```

sarusso commented:
I think you are convoluting things bit: I am not sure if I understand equations 1 and 2, plus the `true_positives_rate` should be never grater than one, and here you get a skyrocketing 13.4...

The `true_positives_rate` should simply be the amount of anomalous events correctly detected by a given model divided by the total amount of actual anomalous events. E.g. 3/5, 0/5, 5/5, but never 6/5 or 8/5.

The first value you already have it: it's just the `true_positives_count`, which in an event-based logic is the amount of anomalous events correctly detected by a given model, indeed.

The second value is instead not yet _explicitly_ computed and I think here is where the issue might be. I propose to include it in the evaluation output as well so that we can clearly see what's going on, and to name it `actual_positives_count`, also to be included in the breakdown.

We can then later add it also for flag-based strategies (where it will be just the total amount of anomalous flags), or you can add it now even if slightly outside the scope of this PR, but that's OK.

The `true_positives_rate` is then given by `true_positives_count / actual_positives_count`, regardless of the strategy.

agataben committed:
Add tests on event evaluation

agataben commented:
I think that a true positives rate greater than 1 is possible: you are not considering cases like this:
```
label                ===============
model output         ===   =   =
```
In this case there is only 1 anomalous event but the model detects 3: the first 3 events count like 1 because points are consecutive, instead the other 2 points count like 2 because they are not consecutive.
So the true positives rate is 3/1.
The last 2 tests added show this.
Later I will add the `actual_positives_count`

sarusso commented:
I see your point, but this is not how event-based evaluation with respect to a given ground truth usually works. For a given actual positive you can have either a true positive or not, but you can't count more than once.

Besides the conceptual "issue", consider also the following example:

```
label               ===============
model 1              ===   =   =
model 2              ==============
```

In this case model 2 would be penalized (as it would count for only one), while model 1 would be rewarded (as it would count for three), but model 2 is actually performing much better than model 1.

The simpler rule is: if any datapoint is detected as anomalous within a given event, then the entire event is marked as  "detected", and I would go with that. So in your example the model should just count for one.

You then _might_ decide to count only partially true positives if the event is not fully detected, but this opens a pandora box on how to effectively count them, and I wouldn't go there.

Agreed on the `actual_positives_count`, let's keep it for later and perhaps in another feature/PR.

agataben committed:
Change variable names

anomalous_events_n ---> events_n
events_by_type_n ---> event_type_counts

agataben committed:
Add new returned value to "_count_anomalous_events()"

"event_time_slots" is a dictionary in which the key is the anomaly type and the value contains timeslots of the anomaly
(a list of start and stop timestamps)

agataben committed:
Build the dictionary "event_time_slots" and add test

agataben committed:
Change the counting strategy of true positives

agataben committed:
Fix error in case of series ending with an anomaly

agataben committed:
Manage series with adjoining anomalies

agataben committed:
Fix error in making the evaluation result dict

agataben committed:
Fix error in counting inserted anomalies

agataben committed:
Add test on event evaluation

agataben commented:
Now anomalies are considered detected if at least one point with an anomaly label is anomalous for the model.  The strategy for counting false positives has remained the same of previous comments.
Below the result of the evaluation on the same dataset as the one used in comments above:
```
evaluation['minmax'] = {
true_positives_count: 7
true_positives_rate: 0.5666666666666667
false_positives_count: 9
false_positives_ratio: 0.0008333333333333335
step_mv_true_positives_count: 6
step_mv_true_positives_rate: 0.9333333333333332
spike_mv_true_positives_count: 1
spike_mv_true_positives_rate: 0.125
}
```
```
evaluation['nhar'] = {
true_positives_count: 7
true_positives_rate: 0.5333333333333333
false_positives_count: 56
false_positives_ratio: 0.005185185185185186
step_mv_true_positives_count: 6
step_mv_true_positives_rate: 0.8
spike_mv_true_positives_count: 1
spike_mv_true_positives_rate: 0.125
}
```
```
evaluation['p_avg'] = {
true_positives_count: 10
true_positives_rate: 0.8333333333333333
false_positives_count: 1597
false_positives_ratio: 0.14787037037037037
step_mv_true_positives_count: 7
step_mv_true_positives_rate: 1.0
spike_mv_true_positives_count: 3
spike_mv_true_positives_rate: 0.625
}
```

agataben committed:
Use function "ensure_full_reproducibility()" to fix the seed

sarusso commented:
I tested this with some models and it seems to work as expected. Merging it!
